### PR TITLE
[SYCL-MLIR]: Simplify creation of dataflow analysis

### PR DESCRIPTION
**Pull request by etiotto**
This patch created a `DataFlowSolverWrapper` class to allow loading required analysis a dataflow analysis might need.
Instead of:
```
    DataFlowSolver solver;
    solver.load<DeadCodeAnalysis>();
    solver.load<SparseConstantPropagation>();
    solver.load<UnderlyingValueAnalysis>();
    solver.load<ReachingDefinitionAnalysis>(aliasAnalysis);
```
the new class allows doing:
```
  DataFlowSolverWrapper solver(aliasAnalysis);
  solver.loadWithRequiredAnalysis<ReachingDefinitionAnalysis>(aliasAnalysis);
```

The requested dataflow analysis (e.g. ReachingDefinitionAnalysis) is then loaded after its required dataflow analysis are loaded. 
In order to use the new framework dataflow analysis need to:
   - inherit from `RequiredDataFlowAnalyses<ReachingDefinitionAnalysis>`
   - implement the static member function `loadRequiredAnalysesImpl`

The `loadRequiredAnalysesImpl` function is responsible for loading required dataflow analysis, for example: 
```
  static void loadRequiredAnalysesImpl(DataFlowSolverWrapper &solver) {
    solver.load<dataflow::DeadCodeAnalysis>();
    solver.load<dataflow::SparseConstantPropagation>();
    solver.load<UnderlyingValueAnalysis>();
  }
```

### CODE REVIEW

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/DataFlowSolverWrapper.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  static void loadRequiredAnalyses(DataFlowSolverWrapper &solver) {
+    Derived::loadRequiredAnalysesImpl(solver);
+  }
+};
+
+/// Provides a way to load into a DataFlowSolver a dataflow analysis, along
+/// with any other dataflow analyses it might require.
+class DataFlowSolverWrapper : public DataFlowSolver {
+public:
+  DataFlowSolverWrapper(AliasAnalysis &aliasAnalysis)
```
Comment (whitneywhtsang) [inline polygeist/include/mlir/Dialect/Polygeist/Analysis/DataFlowSolverWrapper.h:42]: Do we want to allow data flow analyses that don't require AliasAnalysis?

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.h [hunk 2]
```diff
@@ -13,15 +13,103 @@
...
+
+/// An analysis that uses forwarding of values along control-flow and callgraph
+/// edges to determine single underlying values for block arguments.
+class UnderlyingValueAnalysis
+    : public dataflow::SparseDataFlowAnalysis<UnderlyingValueLattice>,
+      public RequiredDataFlowAnalyses<UnderlyingValueAnalysis> {
+  friend class RequiredDataFlowAnalyses;
+
+public:
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(UnderlyingValueAnalysis)
```
Comment (whitneywhtsang) [inline polygeist/include/mlir/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.h:80]: Why is this added? Given that `UnderlyingValueAnalysis` is not in an anonymous namespaces, should we use `MLIR_DECLARE_EXPLICIT_TYPE_ID` and `MLIR_DEFINE_EXPLICIT_TYPE_ID`?
(same for `UnderlyingValueLattice`)

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/DataFlowSolverWrapper.h (hunk 1 above)
Comment (etiotto) [reply polygeist/include/mlir/Dialect/Polygeist/Analysis/DataFlowSolverWrapper.h:42]: Not possible because *some* analysis might depend on a dataflow analysis that requires aliasAnalysis. 
As an example consider `UniformityAnalysis`. It needs the `ReachingDefinitionAnalysis` which needs to be constructed with an alias analysis object. The wrapper calls `loadRequiredAnalysesImpl` on `UniformityAnalysis` (a static member function) and pass a solver. The implementation in of `loadRequiredAnalysesImpl` must load the Reaching def. analysis and so it can only get the alias analysis from the solver.

I'd say for now "this is the way" (feeling like a Mandalorian today ahaha) ... but perhaps we can find a way to make the aliasAnalysis optional.

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.h (hunk 2 above)
Comment (etiotto) [reply polygeist/include/mlir/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.h:80]: Not sure, I guess I should use `MLIR_DECLARE_EXPLICIT_TYPE_ID` ?

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/DataFlowSolverWrapper.h (hunk 1 above)
Comment (whitneywhtsang) [reply polygeist/include/mlir/Dialect/Polygeist/Analysis/DataFlowSolverWrapper.h:42]: ok, we can revisit when we want to use the `DataFlowSolverWrapper` with an analysis that doesn't require `AliasAnalysis`.

**Review by whitneywhtsang — APPROVED** [review]
